common/source/CipherEngine.cpp

```diff
@@ -0,0 +1,218 @@
+//Local Include
+#include "common/CipherEngine.hpp"
+
+// Local Public includes
+#include "common/TraceException.hpp"
+#include "common/DynaLog.hpp"
+
+// Third party includes
+#include <openssl/conf.h>
+#include <openssl/evp.h>
+#include <openssl/rand.h>
+#include <openssl/err.h>
+
+// Standard includes
+#include <iostream>
+#include <exception>
+#include <memory>
+#include <string.h>
+#include <string>
+#include <vector>
+#include <fstream>
+#include <iomanip>
+using namespace std;
+namespace SDMS{
+      
+    void CipherEngine::handleErrors(void)
+    {
+        ERR_print_errors_fp(stderr);
+        abort();
+    } 
+    
+    std::unique_ptr<char[]> CipherEngine::encode64(const unsigned char* input,const int length, LogContext log_context)
+    {
+        // Calculate the padded length based on the input length:
+        // (length + 2) / 3 gives the number of 3-byte blocks (rounded up), multiplied by 4 gives the number of base64 characters required.
+        const int paddedLength = SDMS::CipherEngine::BASE64_ENCODED_BLOCK_SIZE*((length + (SDMS::CipherEngine::BASE64_INPUT_BLOCK_SIZE-1))/SDMS::CipherEngine::BASE64_INPUT_BLOCK_SIZE);
+        auto output = std::make_unique<char[]>(paddedLength+SDMS::CipherEngine::NULL_TERMINATOR_SIZE);
+        std::fill_n(output.get(), paddedLength + SDMS::CipherEngine::NULL_TERMINATOR_SIZE, 0);  // manually zero-initialize
+        const int outputLength = EVP_EncodeBlock(reinterpret_cast<unsigned char*>(output.get()), input, length);
+        if (paddedLength != outputLength) 
+        {  DL_ERROR(log_context, "Output Length (" << outputLength <<") and Predicted Padded Length ("<< paddedLength <<" ) of encoded bytes not equal!"); } 
+        return output;
+    }
+
+    std::unique_ptr<unsigned char[]> CipherEngine::decode64(const char* input,const int length, LogContext log_context) {
```

Comment:
This decode and encode methods can be made const because they are not changing the state of the instance.
```suggestion
    std::unique_ptr<unsigned char[]> CipherEngine::decode64(const char* input,const int length, LogContext log_context) const {
```